fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,252 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.data.metric.MetricColumn;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.PhysicalTable;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+import com.yahoo.bard.webservice.util.Utils;
+
+import com.google.common.collect.Sets;
+
+import org.joda.time.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An implementation of availability which puts metric columns of different availabilities together so that we can
+ * query different metric columns from different availabilities at the same time.
+ * <p>
+ * For example, two availabilities of the following
+ * <pre>
+ * {@code
+ * +---------------+---------------+
+ * | metricColumn1 | metricColumn2 |
+ * +---------------+---------------+
+ * | [1/10]        | [20/30]       |
+ * +---------------+---------------+
+
+ * +---------------+---------------+---------------+
+ * | metricColumn3 | metricColumn4 | metricColumn5 |
+ * +---------------+---------------+---------------+
+ * | [5/15]        | [25/50]       | [90/100]      |
+ * +---------------+---------------+---------------+
+ * }
+ * </pre>
+ * are joined into a metric union availability below
+ * <pre>
+ * {@code
+ * +---------------+---------------+---------------+---------------+---------------+
+ * | metricColumn1 | metricColumn2 | metricColumn3 | metricColumn4 | metricColumn5 |
+ * +---------------+---------------+---------------+---------------+---------------+
+ * | [1/10]        | [20/30]       | [5/15]        | [25/50]       | [90/100]      |
+ * +---------------+---------------+---------------+---------------+---------------+
+ * }
+ * </pre>
+ * Note that the joining availabilities must have completely different set of metric columns.
+ */
```

Comment:
I would reword this and put it higher.  I would say that metric columns available on one availability must not exist on any other availabilities.